enterprise/backend/src/metabase_enterprise/remote_sync/source/git.clj

```diff
@@ -0,0 +1,289 @@
+      (catch Exception e
+        (analytics/inc! :metabase-remote-sync/git-operations-failed analytics-labels)
+        (throw (ex-info (format "Git %s failed: %s" (-> command .getClass .getName) (.getMessage e))
+                        analytics-labels e))))))
+
+(defn- qualify-branch [branch]
+  (if (str/starts-with? branch "refs/heads/")
+    branch
+    (str "refs/heads/" branch)))
+
+(defn fetch!
+  "Call fetch on the repository."
+  [{:keys [^Git git] :as git-source}]
+  (log/info "Fetching repository" {:repo (str git)})
+  (u/prog1 (call-remote-command (.fetch git) git-source))
+  (log/info "Successfully fetched repository" {:repo (str git)}))
+
+(defn- existing-git-repo [^java.io.File dir {:keys [^String token]}]
+  (io/make-parents dir)
+  (try
+    (when (.exists dir)
+      (u/prog1 (Git/open dir)
+        (fetch! {:git <> :token token})))
+    (catch Exception e
+      (log/warnf e "Existing git repo at %s is not configured correctly. Deleting it" dir)
+      (FileUtils/deleteDirectory dir)
+      nil)))
+
+(defn clone-repository!
+  "Clone git repository to a temporary directory using jgit. If the temp directory already exists, use it rather than re-clone.
+  Returns the path to the cloned `jgit.Repository` bare-repo object."
+  [{:keys [^String url ^String token] :as args}]
+  (let [dir (io/file (System/getProperty "java.io.tmpdir") "metabase-git" (-> (str/join ":" [url token]) buddy-hash/sha1 codecs/bytes->hex))
+        existing-git (existing-git-repo dir args)]
+    (if existing-git
+      existing-git
+      (try
+        (log/info "Cloning repository" {:url url :dir dir})
+        (u/prog1 (call-remote-command (-> (Git/cloneRepository)
+                                          (.setDirectory dir)
+                                          (.setURI url)
+                                          (.setBare true)) {:token token})
+          (log/info "Successfully cloned repository" {:dir dir}))
+        (catch Exception e
+          (throw (ex-info (format "Failed to clone git repository: %s" (.getMessage e))
+                          {:url   url
+                           :dir   dir
+                           :error (.getMessage e)} e)))))))
+
+(defn ->commit-id
+  "Returns the full commit ref for a branch or commit-ish string, or nil if not found."
+  ([{:keys [^String commit-ish] :as source}]
+   (->commit-id source commit-ish))
+
+  ([{:keys [^Git git]} ^String commit-ish]
+   (when-let [ref (.resolve (.getRepository git) commit-ish)]
+     (.name ref))))
+
+(defn log
+  "The log of commits on a branch."
+  [{:keys [^Git git] :as source}]
+  (when-let [ref (->commit-id source)]
+    (when-let [branch-id (.resolve (.getRepository git) ref)]
+      (let [log-result (call-command (-> (.log git)
+                                         (.add branch-id)))]
+        (map (fn [^RevCommit commit] {:message      (.getFullMessage commit)
+                                      :author-name  (.getName (.getAuthorIdent commit))
```

Comment:
small nit, i prefer the `..` interop macro for chains of calls like this so
```suggestion
                                      :author-name  (.. commit getAuthorIdent getName)
```
just a bit easier to read. imo, some other examples in this file.